### ISSUE TITLE
Ease communication/API calls between plugins

### ISSUE THREAD

**Issue by mrclay** [Elgg/Elgg#8763]
To avoid the use of things like `function_exists()`, Elgg devs have to bend over backwards to allow their APIs to be used across plugins, e.g. moving logic into views, or using hooks in awkward ways.

E.g. consider options I considered to make a type/subtype likeable:

``` php
// using hook
elgg_register_plugin_hook_handler('likes:is_likable', 'object:mysubtype', 'Elgg\Values::getTrue');

// direct API
if (function_exists('likes_register_type')) {
    likes_register_type('object', 'mysubtype');
}
```

Here the hook-based version we chose is only an option because the API is so simple. Once the API is more complex, it becomes more awkward to stuff it into hooks & handlers and to document it. Worse, static analysis tools can't reason about the API usage; devs can't jump to the API implementation; and the handler is registered even if it's never used.

Basically I think pushing cross-plugin API calls out of normal functions and into hooks and views is a mistake.

One compromise would be a function like:

``` php
// don't care about result or if called or not
elgg_call('likes_register_type', 'object', 'mysubtype');

// need to use return value
$res = elgg_call(MyPlugin\Invokables\SayHello::class, 'Steve'); // calls __invoke()
if ($res->was_called) {
    echo $res->return; // "Hello, Steve"
}
```

**Comment by mrclay** [comment]
A simple convention of exposing a public API object might be the best option:
```php
if (function_exists('likes_api')) {
    likes_api()->registerType('object', 'mysubtype');
}
```
Super easy for devs and tools to investigate/understand this code.

**Comment by mrclay** [comment]
To be fair, the hooks/views-powered options do empower plugins to override the APIs of other plugins. So maybe having `elgg_call` trigger a plugin hook underneath (to select the callback and decorate the result) is more realistic.

**Comment by mrclay** [comment]
Proof of concept #8765

**Comment by mrclay** [comment]
Another idea: formalize a convention where the API is a handler instead of a triggerer. E.g.:

```php
// API provider
elgg_register_plugin_hook_handler('api_call', 'likes:register_type', 'likes_handle_register');

// API user
elgg_trigger_plugin_hook('api_call', 'likes:register_type', null, ['object', 'mysubtype']);
```

The major difference here is when the hooks are called, how often, and if state must be stored. In the current likes design, hooks are only called when needed and no state has to be stored. With the above (and `elgg_call`), the hooks must be called in every init, and the likes plugin must manage the list of registered types somewhere.

**Comment by mrclay** [comment]
We still have time to rework the likes design (and similarly for [discussions](https://github.com/Elgg/Elgg/pull/8761)) for 2.0. What about using config for this?

```php
// register object:file as likeable
$types = (array)elgg_get_config('likes:types');
$types['object:file'] = true;
elgg_set_config('likes:types', $types);
```

or...
```php
elgg_set_config('like:types:object:file', true);
```

Any other ideas?

**Comment by mrclay** [comment]
Another idea toward this goal: #8768 (but 2.0 is frozen)

**Comment by jdalsem** [comment]
Having a lot of plugins developed i still do not think this is going to be worth the trouble
